src/ParserCombinators/Datatypes.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE InstanceSigs #-}
+-- {-# LANGUAGE FlexibleInstances #-}
+-- {-# LANGUAGE UndecidableInstances #-}
+
+module ParserCombinators.Datatypes where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (MonadPlus)
+
+------------------------
+---- ParseErrorType ----
+------------------------
+
+data ParseErrorType i e
+  = EndOfInput
+  | Unexpected i
+  | Expected i i
+  | ExpectedEndOfFile i
+  | CustomError e
+  | Empty
+  deriving (Eq)
+
+instance (Show i, Show e) => Show (ParseErrorType i e) where
```

Comment:
In general it's not a good idea to use `Show` for human readable output. 

It's usually reserved for being derived and meaning "this is how you could write a `ParseErrorType i e` in a haskell source code file".

Prefer having an explicit pretty printing function or even a class if you don't want to remember names.